app/src/behaviors/behavior_tap_dance.c

```diff
@@ -0,0 +1,270 @@
+#include <zmk/behavior.h>
+
+#include <zmk/matrix.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+#include <zmk/events/keycode_state_changed.h>
+#include <zmk/hid.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define ZMK_BHV_TAP_DANCE_MAX_HELD 10
+
+#define ZMK_BHV_TAP_DANCE_POSITION_FREE ULONG_MAX
+
+struct behavior_tap_dance_config {
+    uint32_t tapping_term_ms;
+    int behavior_count;
+    struct zmk_behavior_binding *behaviors;
+};
+
+struct active_tap_dance {
+    // tap dance data
+    int counter;
+    uint32_t position;
+    uint32_t param1;
+    uint32_t param2;
+    bool is_pressed;
+    const struct behavior_tap_dance_config *config;
+    // timer data
+    bool timer_started;
+    bool timer_cancelled;
+    bool timer_timeout;
+    int64_t release_at;
+    struct k_delayed_work release_timer;
+};
+
+struct active_tap_dance active_tap_dances[ZMK_BHV_TAP_DANCE_MAX_HELD] = {};
+
+static struct active_tap_dance *store_tap_dance(uint32_t position,
+                                                const struct behavior_tap_dance_config *config) {
+    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+        struct active_tap_dance *const tap_dance = &active_tap_dances[i];
+        if (tap_dance->position != ZMK_BHV_TAP_DANCE_POSITION_FREE || tap_dance->timer_cancelled) {
+            continue;
+        }
+        tap_dance->counter = 1;
+        tap_dance->position = position;
+        tap_dance->config = config;
+        tap_dance->release_at = 0;
+        tap_dance->timer_started = true;
+        tap_dance->timer_cancelled = false;
+        return tap_dance;
+    }
+    return NULL;
+}
+
+static struct active_tap_dance *find_tap_dance(uint32_t position) {
+    for (int i = 0; i < ZMK_BHV_TAP_DANCE_MAX_HELD; i++) {
+        if (active_tap_dances[i].position == position && !active_tap_dances[i].timer_cancelled) {
+            return &active_tap_dances[i];
+        }
+    }
+    return NULL;
+}
+
+static int stop_timer(struct active_tap_dance *tap_dance) {
+    int timer_cancel_result = k_delayed_work_cancel(&tap_dance->release_timer);
+    if (timer_cancel_result == -EINPROGRESS) {
+        // too late to cancel, we'll let the timer handler clear up.
+        tap_dance->timer_cancelled = true;
+    }
+    return timer_cancel_result;
+}
+
+static void clear_tap_dance(struct active_tap_dance *tap_dance) {
+    LOG_DBG("Clearing Tap Dance");
+    tap_dance->position = ZMK_BHV_TAP_DANCE_POSITION_FREE;
+}
+
+static inline int press_tap_dance_behavior(struct active_tap_dance *tap_dance, int64_t timestamp) {
+    LOG_DBG("Press Tap Dance Behavior");
+    struct zmk_behavior_binding binding = tap_dance->config->behaviors[(tap_dance->counter) - 1];
+    struct zmk_behavior_binding_event event = {
+        .position = tap_dance->position,
+        .timestamp = timestamp,
+    };
+    if (tap_dance->counter <= tap_dance->config->behavior_count) {
```

Comment:
you can also trigger the 'final' tap dance item as soon as the tap dance is tapped N times.

 For example, a tap dance with 3 items would trigger immediately at the 3rd press.